### PR TITLE
Wrap all Node-To-Client queries in a top level query type

### PR DESCRIPTION
**Pull request by DavidEichmann**
This is just a refactoring in preparation for adding a new query: `GetPartialLedgerConfig :: Query blk (PartialLedgerConfig blk)` (I've got a WIP branch where I do that here: [davide/query-ledger-config](https://github.com/input-output-hk/ouroboros-network/tree/davide/query-ledger-config)). You may want to review this commit by commit. The first 2 commits are simple renamings

### CODE REVIEW

File: ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/ByronCompatibility.hs
Comment (nfrisby) [inline ouroboros-consensus-cardano-test/test/Test/Consensus/Cardano/ByronCompatibility.hs:424]: We'll have to check for alignment being busted like it is here.

File: ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Degenerate.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Degenerate.hs:13]: Our `StyleGuide.md` is to leave the paren where it was.

(IIUC `stylish-haskell` on the CI will enforce this, FYI.)

File: ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs:43]: Depending on where our merges fall between node releases, we may be able to add the new top-level queries to `NodeToClientV_9` as well.

**Review by nfrisby — CHANGES_REQUESTED** [review]
Thanks so much for separating these commits nicely. Made my first pass a lot easier.

Most of my comments are style stuff.

There is at least one crucial comment: about backwards-compatible serialization.

File: ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs:43]: See also my comment [here](https://github.com/input-output-hk/ouroboros-network/pull/2975#discussion_r623812670).

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs
Comment (DavidEichmann) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs:104]: @nfrisby I've added this version check, but things are a bit awkeward as the check depends on the `NodeToClientVersion` rather than the `BlockNodeToClientVersion blk`, so I've ended up having to add a `SupportedNetworkProtocolVersion blk` constraint in various places (this is all done in the latest commit). Any thoughts on this or if there is a better way of doing this?

**Comment by DavidEichmann** [comment]
Latest update fixes a few test failures.

**Comment by DavidEichmann** [comment]
Oooops, lost a change due to a local rebase. Latest update fixes a few test failures (for real this time :crossed_fingers:).

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs
Comment (DavidEichmann) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs]: This is resolved in the latest version.

**Comment by DavidEichmann** [comment]
@nfrisby the version checks when serialising `Query` were a bit awkward because they required the `NodeToClientVersion` as opposed to just the `BlockNodeToClientVersion blk` available in `SerialiseNodeToClient`. My solution was just to use a custom function taking the `NodeToClientVersion` for decoding/encoding the `Query` instead of providing a `instance SerialiseNodeToClient blk (SomeSecond Query blk)`. You can see this in the diff linked below.

It's probably best you view the [diff since your last review](https://github.com/input-output-hk/ouroboros-network/compare/0bae6e25aefa79fb433c314356d554d369813244..dc24c93e5007245027638b10c49f0492a1d437ca). This will skip over some changes I've reverted regarding [this comment](https://github.com/input-output-hk/ouroboros-network/pull/3106#discussion_r625066873).

**Comment by DavidEichmann** [comment]
I'm waiting for review. I don't have access to buildkite, so don't know why it's failing on CI.

File: ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs
Comment (nfrisby) [inline ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs:115]: Drop this undent please.

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:190]: Perhaps `codecLocalStateQuery` takes a boolean instead of the version for the same reasons you had mentioned about the awkwardness of importing the right types into the encoder's definition?

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:306]: Hmm. This seems a bit overealous, since line 280 above still has:

```
tStateQueryTracer   :: f (TraceLabelPeer peer (TraceSendRecv (LocalStateQuery blk (Point blk) (Query blk))))
```

We probably want to leave the constraints on `Query` eg so that when we do add new top-level-only-queries this constraint need not change. Does that make sense to you?

I'm guessing this is the result of a find-and-replace etc, so there are probably other signatures and eg other classes that would require the same treatment.

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:306]: IIUC, John Ky has written some of these instances already on his PR #3110

**Comment by nfrisby** [comment]
The build kite failures on 072bdc40c6 are due to `stylish-haskell.`

https://github.com/input-output-hk/ouroboros-network/#formatting lists commands for you to run `stylish-haskell` localling -- be advised: it'll make the changes in place, IIUC.

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs
Comment (DavidEichmann) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:306]: I think I did that originally but it results in a warning:

```
src/Ouroboros/Consensus/Network/NodeToClient.hs:303:16: warning: [-Wsimplifiable-class-constraints]
    • The constraint ‘ShowQuery (Query blk)’ matches
        instance ShowQuery (BlockQuery blk) => ShowQuery (Query blk)
          -- Defined at src/Ouroboros/Consensus/Ledger/Query.hs:50:10
      This makes type inference for inner bindings fragile;
        either use MonoLocalBinds, or simplify it using the instance
    • In the type signature:
        showTracers :: (Show peer,
                        Show (GenTx blk),
                        Show (ApplyTxErr blk),
                        ShowQuery (Query blk),
                        HasHeader blk) =>
                       Tracer m String -> Tracers m peer blk e
    |
303 | showTracers :: ( Show peer
    |                ^^^^^^^^^^^...
```

I could suppress the warning, but it seems simpler to just wait till the new top level query is added. At that point I think GHC will remind me to adjust these constraints. How does that sound?

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:306]: Ah, I see! Harumph.

Yes, I think the extra git commit churn on these lines is probably the right choice instead of adding the warning suppressor to the whole module. (If pragmas could be specified on a per top-level declaration basis I might make the other choice :/)

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:190]: I was suggesting you could change your encoder and decoder to take the `isTopLevelQueryIntroduced` boolean argument, comparable to the `canAcquireTip` flag that `codecLocalStateQuery` already takes.

Maybe that "cuts the knot" with respect to the types; avoiding the awkwardness.

Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:190]: I'll admit I haven't studied the details of the awkwardness you mentioned. I'll do that now.

Comment (DavidEichmann) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:190]: I've been thinking a bit about this. It seems key that there are 2 separate layers of code here: the network layer and the consensus layer. `NodeToClientVersion` corresponds to the network layer version. `BlockNodeToClientVersion blk` corresponds to the consensus layer version.

What I've implicitly done in this PR is assume that the addition of a top level query, `Query blk`, is a change to the network layer. Hence I've added a new `NodeToClientVersion` (version 9). If that's correct, then I can continue on with this. I'd suggest just passing the `NodeToClientVersion` to `codecLocalStateQuery` and doing the version check for both `Query blk` and for  `canAcquireTip` based on that.

But is this assumption really correct? Perhaps the addition of `Query blk` is actually a consensus layer change. Indeed the network layer has always been polymorphic over the query type. If the change is part of the network layer, then that would mean I should add a new version to *all* `BlockNodeToClientVersion blk`s instead of just to `NodeToClientVersion`. Then I'd need to touch all thoughs types again every time I add a new constructor to `Query blk`.

Any thoughts on this would be helpful.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs:120]: My other comment suggests we put this logic off to the next PR, after merging this one. That said, I don't think the results will ever need an additional tag; the query is argument to `encodeResult` only so that we know which type we're serializing.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs:138]: Same as my comment on `queryEncodeResult`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs
Comment (nfrisby) [inline ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs:132]: Same as my comment on `queryEncodeResult`; however, we will need backward compatibility, in general, not just `error` when the negotiated version is less than `V_9` (or whatever the threshold ends up being). We will still need to be able to communicate with "older" nodes.

File: ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs
Comment (nfrisby) [reply ouroboros-consensus/src/Ouroboros/Consensus/Network/NodeToClient.hs:190]: - See my comment on `queryEncodeResult`; we don't _need_ to worry about serialization at all in this PR. It's up to you whether we do or not, though I would prefer to defer it for this PR so we can work out the serialization details on a narrower PR.
- The semantics of `NodeToClientVersion` and `BlockNodeToClientVersion` accumulate in the opposite direction of the dependency: the `NodeToClientVersion` characterizes the behavior of the whole node (not just the network layer), whereas `BlockNodeToClientVersion` characterizes the behavior of a part of the Consensus part of the node. So if the Consensus part changes, then that means the whole node's behavior changes. (Somewhat tangential, but potentially clarifying remark: the network layer is responsible for negotiating the `NodeToClientVersion` with the peer).
- I agree that "the addition of Query blk is actually a consensus layer change", but because of the above point, that still means we'll need to bump `NodeToClientVersion`. We have to whenever we alter the set of possible queries and responses. Here's a historical example of that:

https://github.com/input-output-hk/ouroboros-network/blob/7c8fac89c8ce4c4f9e5d891c7fc1be6613af3750/ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs#L32-L33

- It helps me to think of `NodeToClientVersion` (and its counterpart `NodeToNodeVersion`) as merely the monotonically-increasing natural number that every pair of peers negotiates when connecting. That's its fundamental role and why it needs to reflect every breaking change.
- The evolution of `BlockNodeToClientVersion` versions then is in some sense just a subsequence (not substring) of the evolution of `NodeToClientVersion` versions: its counter corresponds to when the negotiated thing had to change in ways that (only?) `blk`-sensitive Consensus code needs to consider (eg for backwards-compatibility). (That's the number, like 6 in `CardanoNodeToClientVersion6` -- but the RHS of `CardanoNodeToClientVersion6` breaks it down even farther regarding which parts of Consensus are affected).

File: ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs
Comment (nfrisby) [inline ouroboros-consensus-test/src/Test/Util/Serialisation/Roundtrip.hs:333]: `rtWith` seems fine, but so does the pattern in the existing tests. Please make it uniform, either by inlining `rtWith` here and discarding the binding or else change the exists version-sensitive tests to also use `rtWith`.

Edit: The `roundtrip Result` test might not fit with the `rtWith` mold.

File: ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Node/Serialisation.hs
Comment (nfrisby) [inline ouroboros-consensus-mock/src/Ouroboros/Consensus/Mock/Node/Serialisation.hs:9]: I'm not anticipating why we'd need to add UndecidableInstances; new `BlockQuery` and old `Query` are the same thing. Is it really needed here?

Same question applies in `ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Serialisation/SerialiseNodeToClient.hs`

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Generators.hs
Comment (nfrisby) [inline ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Generators.hs:410]: Can we have one `Arbitrary (WithVersion blk (SomeSecond BlockQuery blk)) => Arbitrary (WithVersion blk (SomeSecond Query blk))` instance? I anticipate it belongs somewhere in `ouroboros-consensus-test`, probably next to the `Arbitrary (SomeSecond BlockQuery blk)=> Arbitrary (SomeSecond Query blk)` instance.

File: ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Generators.hs
Comment (nfrisby) [inline ouroboros-consensus-shelley-test/src/Test/Consensus/Shelley/Generators.hs:185]: Same question as in my comment on the corresponding instance in `ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Generators.hs`

**Review by nfrisby — CHANGES_REQUESTED** [review]
I'm Requesting Changes because I don't think the serializers as-is do backwards-compatibility correctly. I made some comments on that, but I also pointed out the higher-level option of simply deferring the serialization changes until the next PR.

I still have to think very carefully each time I consider the relationship between `NodeTo*Version` and `BlockNodeTo*Version`. But I wrote a big comment here that I hope can help clarify their semantics and roles for you.

I hope it makes more sense now, but if not (or even if so), please feel free to join the Consensus planning meeting tomorrow (see eg my calendar) and we can talk about it there.

File: ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs
Comment (nfrisby) [reply ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs:43]: > IIUC, nothing needs to change here yet.

I think I came to that same conclusion, eg here https://github.com/input-output-hk/ouroboros-network/pull/3106#discussion_r629762791

File: ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs
Comment (DavidEichmann) [reply ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs:52]: I assume this is no longer applicable.

File: ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Generators.hs
Comment (nfrisby) [reply ouroboros-consensus-cardano-test/src/Test/Consensus/Cardano/Generators.hs]: Hmmm. I find overlapping instances tend to add more confusion than they're worth. In general, I prefer explicitness for the sake of transparency.

In this case, I'm hesitant to make the existing overlapping instance _more_ general. I'm guessing the earlier logic was that there are only two `ShelleyNodeToClientVersions` and the difference between them was the addition of one query. So most of the instances were trivial, hence the overlapping was used for that.

But with the new more general instance, this would apply to _any_ version type, even if it were rich. That's not a risk right now, since we already have instances for all the other types (since there was no such general instance), but this opens the door for us to accidentally rely on this general instance for a future version type.

However, that said, this is for `Arbitrary`. So the greatest risk is that we'd see test failures for mismatched versions and values. Once we debug the failure to see the cause, we can add the corresponding override instance. So I'm optimistic here. Let's go with what you have here.

**Review by nfrisby — APPROVED** [review]
LGTM! Looking forward to the new queries.

Please rebase on `origin/master` before telling `bors` to merge.

**Comment by DavidEichmann** [comment]
bors merge

**Comment by DavidEichmann** [comment]
@nfrisby Thanks for keeping up the review :-D I know it's been a long process.

**Comment by nfrisby** [comment]
> Thanks for keeping up the review :-D I know it's been a long process.

Not at all! It's a non-trivial codebase, and we're spinning multiple plates :)

**Pull request merged**
